test/test-template-engine.js

```diff
@@ -729,6 +755,107 @@ describe('html updaters', () => {
+      assert(container.querySelector('#target').textContent === '');
+    };
+
+    it('can change from undefined content to null content', () => run(toUndefinedContent, toNullContent));
+    it('can change from undefined content to text content', () => run(toUndefinedContent, toTextContent));
+    it('can change from undefined content to array content', () => run(toUndefinedContent, toArrayContent));
+    it('can change from undefined content to map', () => run(toUndefinedContent, toMap));
+    it('can change from undefined content to unsafe html', () => run(toUndefinedContent, toUnsafeHTML));
+
+    it('can change from null content to undefined content', () => run(toNullContent, toUndefinedContent));
+    it('can change from null content to text content', () => run(toNullContent, toTextContent));
+    it('can change from null content to array content', () => run(toNullContent, toArrayContent));
+    it('can change from null content to map', () => run(toNullContent, toMap));
+    it('can change from null content to unsafe html', () => run(toNullContent, toUnsafeHTML));
+
+    it('can change from text content to undefined content', () => run(toTextContent, toUndefinedContent));
+    it('can change from text content to null content', () => run(toTextContent, toNullContent));
+    it('can change from text content to array content', () => run(toTextContent, toArrayContent));
+    it('can change from text content to map', () => run(toTextContent, toMap));
+    it('can change from text content to unsafe html', () => run(toTextContent, toUnsafeHTML));
+
+    it('can change from array content to undefined content', () => run(toArrayContent, toUndefinedContent));
+    it('can change from array content to null content', () => run(toArrayContent, toNullContent));
+    it('can change from array content to text content', () => run(toArrayContent, toTextContent));
+    it('can change from array content to map', () => run(toArrayContent, toMap));
+    it('can change from array content to unsafe html', () => run(toArrayContent, toUnsafeHTML));
+
+    it('can change from map to undefined content', () => run(toMap, toUndefinedContent));
+    it('can change from map to null content', () => run(toMap, toNullContent));
+    it('can change from map to text content', () => run(toMap, toTextContent));
+    it('can change from map to array content', () => run(toMap, toArrayContent));
+    it('can change from map to unsafe html', () => run(toMap, toUnsafeHTML));
+
+    it('can change from unsafeHtml to undefined content', () => run(toUnsafeHTML, toUndefinedContent));
+    it('can change from unsafeHtml to null content', () => run(toUnsafeHTML, toNullContent));
+    it('can change from unsafeHtml to text content', () => run(toUnsafeHTML, toTextContent));
+    it('can change from unsafeHtml to array content', () => run(toUnsafeHTML, toArrayContent));
+    it('can change from unsafeHtml to map', () => run(toUnsafeHTML, toMap));
```

Comment:
Took me longer than I‘d like to admit to figure out how to write _expressive_ tests for this. I settled on this format, which feels pretty readable to me. Pretty happy with how this turned out – I ended up catching / fixing quite a few edge-case bugs. For example, previously if you bound `unsafeHTML(/* stuff */)` to some content and then bound `null` to that same content — the prior html didn’t actually get cleared for [reasons].

This should (at least minimally) test every possible content transition that we have today.